core/shared/src/main/scala/clue/ApolloClient.scala

```diff
@@ -1,119 +1,469 @@
+  ) extends State[F, CP](PersistentClientStatus.Initialized)
+  // Reestablishing = We are in the process of reconnecting + reinitializing after a low level error/close, but we haven't connected yet.
+  final case class Reestablishing[F[_]](
+    initPayloadF:  F[Map[String, Json]],
+    subscriptions: Map[String, Emitter[F]],
+    connectLatch:  Latch[F],
+    initLatch:     Latch[F]
+  ) extends State[F, Nothing](PersistentClientStatus.Connecting)
 }
 
-abstract class ApolloClient[F[_]: ConcurrentEffect: Timer: Logger, S, CP, CE](
-  uri:  Uri,
-  name: String
-) extends GraphQLPersistentClient[F, S, CP, CE] {
-  private val LogPrefix = s"[clue.ApolloClient][${if (name.isEmpty) uri else name}]"
+class ApolloClient[F[_], S, CP, CE](
+  uri:                  Uri,
+  reconnectionStrategy: ReconnectionStrategy[CE],
+  state:                Ref[F, State[F, CP]],
+  connectionStatus:     SignallingRef[F, PersistentClientStatus]
+)(implicit
+  F:                    ConcurrentEffect[F],
+  backend:              PersistentBackend[F, CP, CE],
+  timer:                Timer[F],
+  logger:               Logger[F]
+) extends PersistentStreamingClient[F, S, CP, CE]
+    with PersistentBackendHandler[F, CE] {
+  import State._
+
+  // Transition FSM state and execute an action.
+  private def stateModify(f: State[F, CP] => (State[F, CP], F[Unit])): F[Unit] =
+    state
+      .modify { oldState =>
+        val (newState, action) = f(oldState)
+        newState -> ((oldState, newState, action))
+      }
+      .flatMap { case (oldState, newState, action) =>
+        s"State Modified [$oldState] ==> [$newState]".debugF >>
+          connectionStatus.set(newState.status) >>
+          action
+      }
+
+  // <ApolloClient>
+  override def status: F[PersistentClientStatus] =
+    connectionStatus.get
 
-  private val F = implicitly[ConcurrentEffect[F]]
+  override def statusStream: fs2.Stream[F, PersistentClientStatus] =
+    connectionStatus.discrete
 
-  val connectionStatus: SignallingRef[F, StreamingClientStatus]
-  protected val subscriptions: Ref[F, Map[String, Emitter[F]]]
-  protected val firstInitInvoked: Deferred[F, Unit]
-  protected val connectionRef: Ref[F, ApolloClient.Connection[F, CP]]
-  protected val connectionAttempt: Ref[F, Int]
+  override def connect(): F[Unit] = {
+    val warn = "connect() called while already connected or attempting to connect.".warnF
 
-  private val getConnectionDef: F[ApolloClient.ConnectionDef[F, CP]] =
-    firstInitInvoked.get >>
-      connectionRef.get.flatMap(_ match {
-        case None                => F.raiseError(new Exception("Client is disconnected"))
-        case Some(connectionDef) => F.pure(connectionDef)
-      })
+    Latch[F].flatMap { newLatch =>
+      stateModify {
+        case Disconnected                              => Connecting(newLatch) -> doConnect(newLatch)
+        case s @ Connecting(latch)                     => s                    -> (warn >> latch.get.rethrow)
```

Comment:
Any reason to use `>>` over `*>`?

Reply:
In general, I prefer `>>` when things should be run in sequence, making sure we are in the `Monad` world and not in an `Applicative`. Also, the parameter to `>>` is lazily evaluated, while `>*`'s is eager, which may be an issue when there's recursion involved.

Fabio explains it better than me here: https://gitter.im/typelevel/cats?at=5bbf6d49bbdc0b250523988c